src/common/test-jsonfds.c

```diff
@@ -0,0 +1,905 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define _GNU_SOURCE
+
+#include "cockpitcontrolmessages.h"
+#include "cockpitjsonprint.h"
+#include "cockpitmemfdread.h"
+#include "cockpittest.h"
+
+#include <gio/gunixfdmessage.h>
+#include <gio/gunixcredentialsmessage.h>
+
+#include <fcntl.h>
+#include <sys/socket.h>
+#include <sys/mman.h>
+#include <stdio.h>
+#include <unistd.h>
+
+/* --- testing of printing --- */
+
+typedef struct
+{
+  char buffer[1024];
+  FILE *stream;
+
+  char expected_buffer[1024];
+  gchar *expected;
```

Comment:
I find that naming a bit hard to follow when reading the code. `helpers->expected` pretty well always points to a `\0`, so it's not really the "expected" output (that is `expected_buffer`), more like an "expected_end".

Reply:
Reasonable proposal.  I'll change it.